zebra-network/src/peer/connection.rs

```diff
@@ -677,33 +713,36 @@ where
                     .send(Message::GetHeaders { known_blocks, stop })
                     .await
                 {
-                    Ok(()) => Ok((
+                    Ok(()) => Ok(
                         AwaitingResponse {
                             handler: Handler::FindHeaders,
                             tx,
                             span,
                         },
-                        None,
-                    )),
+                    ),
                     Err(e) => Err((e, tx)),
                 }
             }
             (AwaitingRequest, MempoolTransactions) => {
                 match self.peer_tx.send(Message::Mempool).await {
-                    Ok(()) => Ok((
+                    Ok(()) => Ok(
                         AwaitingResponse {
                             handler: Handler::MempoolTransactions,
                             tx,
                             span,
                         },
-                        None,
-                    )),
+                    ),
                     Err(e) => Err((e, tx)),
                 }
             }
             (AwaitingRequest, PushTransaction(transaction)) => {
                 match self.peer_tx.send(Message::Tx(transaction)).await {
-                    Ok(()) => Ok((AwaitingRequest, Some(tx))),
+                    Ok(()) => {
+                        // Since we're not waiting for further messages, we need to
+                        // send a response before dropping tx.
+                        let _ = tx.send(Ok(Response::Nil));
+                        Ok(AwaitingRequest)
```

Comment:
Copying this code feels like an anti-pattern - and it's a reversion of recent changes to this file.
Can we handle the oneshot request Nil responses in a different way?

For example, the peer routing for the Advertise* requests is handled by passing them to different functions:
https://github.com/ZcashFoundation/zebra/blob/main/zebra-network/src/peer_set/set.rs#L469

Reply:
making it so I didn't have to handle propagating the `tx` back out of these cases was pretty critical for removing the need to handle cases that shouldn't ever exist in practice. We can factor this out into a function but I don't see the point, it's one line of code